README.md
<div align="center">

# relative-time

[![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/tbidne/relative-time?include_prereleases&sort=semver)](https://github.com/tbidne/relative-time/releases/)
[![MIT](https://img.shields.io/github/license/tbidne/relative-time?color=blue)](https://opensource.org/licenses/MIT)

[![nix](https://img.shields.io/github/workflow/status/tbidne/relative-time/nix/main?label=nix%209.2.2&&logo=nixos&logoColor=85c5e7&labelColor=2f353c)](https://github.com/tbidne/relative-time/actions/workflows/nix_ci.yaml)
[![stack](https://img.shields.io/github/workflow/status/tbidne/relative-time/stack/main?label=stack%2019.4&logoColor=white&labelColor=2f353c)](https://github.com/tbidne/relative-time/actions/workflows/stack_ci.yaml)
[![style](https://img.shields.io/github/workflow/status/tbidne/relative-time/style/main?label=style&logoColor=white&labelColor=2f353c)](https://github.com/tbidne/relative-time/actions/workflows/style_ci.yaml)

[![8.10.7](https://img.shields.io/github/workflow/status/tbidne/relative-time/8.10.7/main?label=8.10.7&logo=haskell&logoColor=904d8c&labelColor=2f353c)](https://github.com/tbidne/relative-time/actions/workflows/cabal_ci.yaml)
[![9.0.2](https://img.shields.io/github/workflow/status/tbidne/relative-time/9.0.2/main?label=9.0.2&logo=haskell&logoColor=904d8c&labelColor=2f353c)](https://github.com/tbidne/relative-time/actions/workflows/ghc_9-0-2.yaml)
[![9.2.2](https://img.shields.io/github/workflow/status/tbidne/relative-time/9.2.2/main?label=9.2.2&logo=haskell&logoColor=904d8c&labelColor=2f353c)](https://github.com/tbidne/relative-time/actions/workflows/ghc_9-2-2.yaml)

</div>


CHANGELOG.md
# Revision history for timerep

## 0.1 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/unit/Main.hs
-- | Runs unit tests.
module Main (main) where

import Control.Monad qualified as M
import Data.Proxy (Proxy (..))
import System.Environment qualified as Env
import System.Exit qualified as SysEx
import Test.Tasty qualified as Tasty
import Test.Tasty.Options (OptionDescription (..))
import Text.Read qualified as TR
import Unit.Data.Time.Relative qualified as Relative
import Unit.MaxRuns (MaxRuns (..))

-- | Entry point for unit tests.
main :: IO ()
main = do
  maxRuns <-
    Env.lookupEnv "MAX_RUNS" >>= \case
      Nothing -> pure 100
      Just mr -> case parseMaxRuns mr of
        Nothing -> SysEx.die $ "*** MAX_RUNS is not a non-negative integer: " <> mr
        Just x -> pure $ fromIntegral @Int x

  let maxRunProps = Tasty.localOption (MkMaxRuns maxRuns) Relative.props

  Tasty.defaultMainWithIngredients ingredients $
    Tasty.testGroup "Unit tests" [maxRunProps]
  where
    parseMaxRuns = M.mfilter (> 0) . TR.readMaybe
    ingredients = Tasty.includingOptions [Option @MaxRuns Proxy] : Tasty.defaultIngredients


test/doctest/Main.hs
module Main (main) where

import System.Environment qualified as Env
import Test.DocTest qualified as DocTest
import Prelude

main :: IO ()
main = do
  shouldRun <- Env.lookupEnv "RUN_DOCTEST"
  case shouldRun of
    Just "true" -> DocTest.doctest args
    _ -> putStrLn "*** Doctests Disabled ***"
  where
    args = files <> exts

files :: [String]
files =
  [ "-isrc",
    "src/Data/Time/Relative.hs"
  ]

exts :: [String]
exts =
  [ "-XApplicativeDo",
    "-XDeriveAnyClass",
    "-XDeriveDataTypeable",
    "-XDeriveGeneric",
    "-XDerivingStrategies",
    "-XDerivingVia",
    "-XImportQualifiedPost",
    "-XLambdaCase",
    "-XMultiParamTypeClasses",
    "-XNumericUnderscores",
    "-XOverloadedStrings",
    "-XTypeApplications"
  ]


test/unit/Unit/MaxRuns.hs
-- | Provides the 'MaxRuns' type.
--
-- @since 0.1
module Unit.MaxRuns
  ( MaxRuns (..),
  )
where

import Data.Tagged (Tagged (..))
import Hedgehog (TestLimit)
import Test.Tasty.Options (IsOption (..))
import Text.Read qualified as TR

-- | Sets the maximum successful runs for each test.
--
-- @since 0.1
newtype MaxRuns = MkMaxRuns TestLimit
  deriving stock (Show)
  deriving (Num) via TestLimit

-- | @since 0.1
instance IsOption MaxRuns where
  defaultValue = MkMaxRuns 100
  parseValue = readLimit
  optionName = Tagged "max-runs"
  optionHelp = Tagged "The maximum number of runs for each test."

readLimit :: [Char] -> Maybe MaxRuns
readLimit = fmap (fromIntegral @Int) . TR.readMaybe


src/Data/TimeRep.hs
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE ImportQualifiedPost #-}
{-# LANGUAGE NumericUnderscores #-}

-- | Provides the 'TimeRep' type and related functions for representing
-- time.
module Data.TimeRep
  ( -- * Type
    TimeRep (..),

    -- * Conversions
    toSeconds,
    fromSeconds,
    fromString,

    -- * Formatting
    formatTimeRep,
    formatTime,
  )
where

import Control.Applicative (Alternative (..))
import Control.DeepSeq (NFData)
import Data.Data (Data)
import Data.Foldable (foldl')
import Data.List qualified as L
import GHC.Generics (Generic)
import GHC.Natural (Natural)
import GHC.Read qualified as GRead
import Numeric.Algebra (AMonoid (..), ASemigroup (..))
import Text.ParserCombinators.ReadP qualified as RP
import Text.ParserCombinators.ReadPrec (ReadPrec, (+++))
import Text.ParserCombinators.ReadPrec qualified as RPC
import Text.Read (Read (..))
import Text.Read.Lex (Lexeme (..))

-- | Represents a relative time with second granularity. This is primarily
-- intended to be used with user supplied numeric values for convenience.
--
-- For example, suppose an application takes an argument representing a
-- timeout. Asking for a natural number representing seconds is reasonable,
-- but it has low UX if there is any chance this timeout could be somewhat
-- large e.g. over an hour.
--
-- Instead, we can allow the user supply a "time string" like "1d2h3m4s".
-- 'fromString' will parse this into 'TimeRep', and then the application can
-- either convert this into seconds or keep as a 'TimeRep' as needed.
--
-- @since 0.1
data TimeRep = MkTimeRep
  { -- | @since 0.1
    days :: !Natural,
    -- | @since 0.1
    hours :: !Natural,
    -- | @since 0.1
    minutes :: !Natural,
    -- | @since 0.1
    seconds :: !Natural
  }
  deriving stock
    ( -- | @since 0.1
      Eq,
      -- | @since 0.1
      Data,
      -- | @since 0.1
      Generic,
      -- | @since 0.1
      Show
    )
  deriving anyclass
    ( -- | @since 0.1
      NFData
    )

-- @since 0.1
instance Ord TimeRep where
  x <= y = toSeconds x <= toSeconds y

-- @since 0.1
instance Read TimeRep where
  readPrec = readRecord +++ readSeconds +++ readTimeStr

-- @since 0.1
instance ASemigroup TimeRep where
  MkTimeRep d h m s .+. MkTimeRep d' h' m' s' =
    MkTimeRep (d + d') (h + h') (m + m') (s + s')

-- @since 0.1
instance AMonoid TimeRep where
  zero = MkTimeRep 0 0 0 0

-- | Transforms a 'TimeRep' into 'Natural' seconds.
--
-- @since 0.1
toSeconds :: TimeRep -> Natural
toSeconds (MkTimeRep d h m s) =
  d * secondsInDay
    + h * secondsInHour
    + m * secondsInMinute
    + s

-- | Transforms 'Natural' seconds into a 'TimeRep'.
--
-- @since 0.1
fromSeconds :: Natural -> TimeRep
fromSeconds seconds' = MkTimeRep d h m s
  where
    (d, daysRem) = seconds' `quotRem` secondsInDay
    (h, hoursRem) = daysRem `quotRem` secondsInHour
    (m, s) = hoursRem `quotRem` secondsInMinute

-- | Converts a 'String' into a 'TimeRep'.
--
-- @since 0.1
fromString :: String -> Either String TimeRep
fromString str = case RPC.readPrec_to_S read' RPC.minPrec str of
  [(y, "")] -> Right y
  _ -> Left $ "Could not read TimeRep from: " <> str
  where
    read' = readTimeStr +++ readSeconds <* RPC.lift RP.skipSpaces

-- | Formats a 'TimeRep' to 'String'.
--
-- @since 0.1
formatTimeRep :: TimeRep -> String
formatTimeRep (MkTimeRep 0 0 0 0) = "0 seconds"
formatTimeRep (MkTimeRep d h m s) = L.intercalate ", " vals
  where
    f acc (n, units)
      | n == 0 = acc
      | otherwise = pluralize n units : acc
    vals = foldl' f [] [(s, " second"), (m, " minute"), (h, " hour"), (d, " day")]

-- | For \(n \ge 0\) seconds, returns a 'String' description of the days, hours,
-- minutes and seconds.
--
-- @since 0.1
formatTime :: Natural -> String
formatTime = formatTimeRep . fromSeconds

pluralize :: Natural -> String -> String
pluralize n txt
  | n == 1 = valUnit
  | otherwise = valUnit <> "s"
  where
    valUnit = show n <> txt

secondsInDay :: Natural
secondsInDay = 86_400

secondsInHour :: Natural
secondsInHour = 3_600

secondsInMinute :: Natural
secondsInMinute = 60

readRecord :: ReadPrec TimeRep
readRecord = GRead.parens $
  RPC.prec 11 $ do
    GRead.expectP (Ident "MkTimeRep")
    GRead.expectP (Punc "{")
    d <- GRead.readField "days" (RPC.reset GRead.readPrec)
    GRead.expectP (Punc ",")
    h <- GRead.readField "hours" (RPC.reset GRead.readPrec)
    GRead.expectP (Punc ",")
    m <- GRead.readField "minutes" (RPC.reset GRead.readPrec)
    GRead.expectP (Punc ",")
    s <- GRead.readField "seconds" (RPC.reset GRead.readPrec)
    GRead.expectP (Punc "}")
    pure $ MkTimeRep d h m s

readSeconds :: ReadPrec TimeRep
readSeconds = fromSeconds <$> readPrec

readTimeStr :: ReadPrec TimeRep
readTimeStr =
  MkTimeRep
    <$> readTimeOrZero 'd'
    <*> readTimeOrZero 'h'
    <*> readTimeOrZero 'm'
    <*> readTimeOrZero 's'

readTimeOrZero :: Char -> ReadPrec Natural
readTimeOrZero c =
  readTimeWithUnit c <|> pure 0

readTimeWithUnit :: Char -> ReadPrec Natural
readTimeWithUnit c = readPrec <* RPC.lift (RP.char c)


src/Data/Time/Relative.hs
-- | Provides the 'RelativeTime' type and related functions for representing
-- time.
--
-- @since 0.1
module Data.Time.Relative
  ( -- * Type
    RelativeTime (..),

    -- * Operations
    -- $operations
    normalize,

    -- * Conversions
    toSeconds,
    fromSeconds,
    fromString,

    -- * Formatting
    formatRelativeTime,
    formatSeconds,
  )
where

import Control.Applicative (Alternative (..))
import Control.DeepSeq (NFData)
import Data.Data (Data)
import Data.Foldable (foldl')
import Data.List qualified as L
import GHC.Generics (Generic)
import GHC.Natural (Natural)
import GHC.Read qualified as GRead
import Numeric.Algebra (AMonoid (..), ASemigroup (..), Semimodule (..))
import Text.ParserCombinators.ReadP qualified as RP
import Text.ParserCombinators.ReadPrec (ReadPrec, (+++))
import Text.ParserCombinators.ReadPrec qualified as RPC
import Text.Read (Read (..))
import Text.Read.Lex (Lexeme (..))

-- | Represents a relative time with second precision. This is primarily
-- intended to be used with user supplied numeric values for convenience.
--
-- For example, suppose an application takes an argument representing a
-- timeout. Asking for a natural number representing seconds is reasonable,
-- but it has low UX if there is any chance this timeout could be somewhat
-- large e.g. over an hour.
--
-- Instead, we can allow the user to supply a "time string" like "1d2h3m4s".
-- 'fromString' will parse this into 'RelativeTime', and then the application can
-- either convert this into seconds or keep as a 'RelativeTime' as needed.
--
-- ==== __Instances__
--
-- * 'Eq'/'Ord': Terms are converted first to seconds then compared i.e. we
--   declare an equivalence class in terms of the "total time" represented.
-- * 'Read': Parses the same strings as 'fromString'. Additionally, we also
--   parse the output of 'Show' i.e. the derived instance.
--
-- >>> read @RelativeTime "MkRelativeTime {days = 1, hours = 2, minutes = 3, seconds = 4}"
-- MkRelativeTime {days = 1, hours = 2, minutes = 3, seconds = 4}
--
-- @since 0.1
data RelativeTime = MkRelativeTime
  { -- | @since 0.1
    days :: !Natural,
    -- | @since 0.1
    hours :: !Natural,
    -- | @since 0.1
    minutes :: !Natural,
    -- | @since 0.1
    seconds :: !Natural
  }
  deriving stock
    ( -- | @since 0.1
      Eq,
      -- | @since 0.1
      Data,
      -- | @since 0.1
      Generic,
      -- | @since 0.1
      Show
    )
  deriving anyclass
    ( -- | @since 0.1
      NFData
    )

-- | @since 0.1
instance Ord RelativeTime where
  x <= y = toSeconds x <= toSeconds y

-- | @since 0.1
instance Read RelativeTime where
  readPrec = readRecord +++ readSeconds +++ readTimeStr

-- | @since 0.1
instance ASemigroup RelativeTime where
  MkRelativeTime d h m s .+. MkRelativeTime d' h' m' s' =
    normalize $ MkRelativeTime (d + d') (h + h') (m + m') (s + s')

-- | @since 0.1
instance AMonoid RelativeTime where
  zero = MkRelativeTime 0 0 0 0

-- | @since 0.1
instance Semimodule RelativeTime Natural where
  MkRelativeTime d h m s .* k =
    normalize $ MkRelativeTime (d * k) (h * k) (m * k) (s * k)

-- $operations
-- Operations on 'RelativeTime'. In addition to the unary 'normalize', we also
-- have instances from @algebra-simple@:
--
-- >>> let t1 = MkRelativeTime 1 2 3 4
-- >>> let t2 = MkRelativeTime 2 3 4 5
-- >>> t1 .+. t2
-- MkRelativeTime {days = 3, hours = 5, minutes = 7, seconds = 9}
--
-- >>> t1 .+. zero
-- MkRelativeTime {days = 1, hours = 2, minutes = 3, seconds = 4}
--
-- >>> t1 .* 2
-- MkRelativeTime {days = 2, hours = 4, minutes = 6, seconds = 8}
--
-- These operations are 'normalize'd.

-- | Transforms the 'RelativeTime' into "canonical" form i.e. if we can reduce
-- some unit (e.g. 75 seconds), then we do.
--
-- ==== __Examples__
--
-- >>> normalize $ MkRelativeTime 0 0 0 75
-- MkRelativeTime {days = 0, hours = 0, minutes = 1, seconds = 15}
--
-- >>> normalize $ MkRelativeTime 2 23 59 61
-- MkRelativeTime {days = 3, hours = 0, minutes = 0, seconds = 1}
--
-- @since 0.1
normalize :: RelativeTime -> RelativeTime
normalize = fromSeconds . toSeconds

-- | Transforms a 'RelativeTime' into 'Natural' seconds.
--
-- ==== __Examples__
--
-- >>> toSeconds $ MkRelativeTime 0 1 1 1
-- 3661
--
-- >>> toSeconds $ MkRelativeTime 1 2 3 4
-- 93784
--
-- @since 0.1
toSeconds :: RelativeTime -> Natural
toSeconds (MkRelativeTime d h m s) =
  d * secondsInDay
    + h * secondsInHour
    + m * secondsInMinute
    + s

-- | Transforms 'Natural' seconds into a 'RelativeTime'.
--
-- ==== __Examples__
--
-- >>> fromSeconds 3661
-- MkRelativeTime {days = 0, hours = 1, minutes = 1, seconds = 1}
--
-- >>> fromSeconds 93784
-- MkRelativeTime {days = 1, hours = 2, minutes = 3, seconds = 4}
--
-- @since 0.1
fromSeconds :: Natural -> RelativeTime
fromSeconds seconds' = MkRelativeTime d h m s
  where
    (d, daysRem) = seconds' `quotRem` secondsInDay
    (h, hoursRem) = daysRem `quotRem` secondsInHour
    (m, s) = hoursRem `quotRem` secondsInMinute

-- | Converts a string into a 'RelativeTime'. Converts either a
-- "time string" e.g. "1d2h3m4s" or numeric literal (interpreted
-- as seconds).
--
-- ==== __Examples__
--
-- >>> fromString "1d2h3m4s"
-- Right (MkRelativeTime {days = 1, hours = 2, minutes = 3, seconds = 4})
--
-- >>> fromString "1h1s"
-- Right (MkRelativeTime {days = 0, hours = 1, minutes = 0, seconds = 1})
--
-- >>> fromString "0h15s"
-- Right (MkRelativeTime {days = 0, hours = 0, minutes = 0, seconds = 15})
--
-- >>> fromString "3601"
-- Right (MkRelativeTime {days = 0, hours = 1, minutes = 0, seconds = 1})
--
-- >>> fromString ""
-- Right (MkRelativeTime {days = 0, hours = 0, minutes = 0, seconds = 0})
--
-- >>> fromString "1s1h"
-- Left "Could not read RelativeTime from: 1s1h"
--
-- >>> fromString "cat"
-- Left "Could not read RelativeTime from: cat"
--
-- @since 0.1
fromString :: String -> Either String RelativeTime
fromString str =
  case [x | (x, "") <- RPC.readPrec_to_S read' RPC.minPrec str] of
    [y] -> Right y
    _ -> Left $ "Could not read RelativeTime from: " <> str
  where
    read' = readTimeStr +++ readSeconds <* RPC.lift RP.skipSpaces

-- | Formats a 'RelativeTime' to string.
--
-- ==== __Examples__
--
-- >>> formatRelativeTime $ MkRelativeTime 1 2 0 3
-- "1 day, 2 hours, 3 seconds"
--
-- @since 0.1
formatRelativeTime :: RelativeTime -> String
formatRelativeTime (MkRelativeTime 0 0 0 0) = "0 seconds"
formatRelativeTime (MkRelativeTime d h m s) = L.intercalate ", " vals
  where
    f acc (n, units)
      | n == 0 = acc
      | otherwise = pluralize n units : acc
    vals = foldl' f [] [(s, " second"), (m, " minute"), (h, " hour"), (d, " day")]

-- | For \(n \ge 0\) seconds, returns a string description of the days, hours,
-- minutes and seconds.
--
-- ==== __Examples__
--
-- >>> formatSeconds 3623
-- "1 hour, 23 seconds"
--
--
-- @since 0.1
formatSeconds :: Natural -> String
formatSeconds = formatRelativeTime . fromSeconds

pluralize :: Natural -> String -> String
pluralize n txt
  | n == 1 = valUnit
  | otherwise = valUnit <> "s"
  where
    valUnit = show n <> txt

secondsInDay :: Natural
secondsInDay = 86_400

secondsInHour :: Natural
secondsInHour = 3_600

secondsInMinute :: Natural
secondsInMinute = 60

readRecord :: ReadPrec RelativeTime
readRecord = GRead.parens $
  RPC.prec 11 $ do
    GRead.expectP (Ident "MkRelativeTime")
    GRead.expectP (Punc "{")
    d <- GRead.readField "days" (RPC.reset GRead.readPrec)
    GRead.expectP (Punc ",")
    h <- GRead.readField "hours" (RPC.reset GRead.readPrec)
    GRead.expectP (Punc ",")
    m <- GRead.readField "minutes" (RPC.reset GRead.readPrec)
    GRead.expectP (Punc ",")
    s <- GRead.readField "seconds" (RPC.reset GRead.readPrec)
    GRead.expectP (Punc "}")
    pure $ MkRelativeTime d h m s

readSeconds :: ReadPrec RelativeTime
readSeconds = fromSeconds <$> readPrec

readTimeStr :: ReadPrec RelativeTime
readTimeStr =
  MkRelativeTime
    <$> readTimeOrZero 'd'
    <*> readTimeOrZero 'h'
    <*> readTimeOrZero 'm'
    <*> readTimeOrZero 's'

readTimeOrZero :: Char -> ReadPrec Natural
readTimeOrZero c =
  readTimeWithUnit c <|> pure 0

readTimeWithUnit :: Char -> ReadPrec Natural
readTimeWithUnit c = readPrec <* RPC.lift (RP.char c)
